src/Processors/QueryPlan/ReadFromRemote.cpp

```diff
@@ -431,12 +439,25 @@ void ReadFromRemote::addLazyPipe(Pipes & pipes, const ClusterProxy::SelectStream
     if (filter_actions_dag)
         pushed_down_filters = std::make_shared<const ActionsDAG>(filter_actions_dag->clone());
 
+    const StorageID resolved_id = context->resolveStorageID(main_table);
+    const StoragePtr storage = DatabaseCatalog::instance().tryGetTable(resolved_id, context);
+    if (!storage)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Storage with id {} not found", resolved_id);
```

Comment:
Should this be `LOGICAL_ERROR`? And what if the underlying table had been removed?
Also isn't the `main_table` will be empty in case of `select * from remote remote()`?

Reply:
Changed the error to `UNKNOWN_TABLE`.

Reply:
> Also isn't the main_table will be empty in case of `select * from remote remote()`?

The `addLazyPipe` function will not be called for tables that don't exist locally:
https://github.com/ClickHouse/ClickHouse/blob/master/src/Interpreters/ClusterProxy/SelectStreamFactory.cpp#L230 (`lazy = false` is passed here).

By the way, I've changed the code to use `main_table` only  if `shard.main_table` is empty, in line with this: https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromRemote.h#L23